src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -733,32 +733,62 @@ private async Task AcquireWriteLockAsync(CancellationToken cancellationToken)
 
         private async Task SendSettingsAckAsync()
         {
-            await StartWriteAsync(FrameHeader.Size).ConfigureAwait(false);
-            WriteFrameHeader(new FrameHeader(0, FrameType.Settings, FrameFlags.Ack, 0));
+            try
+            {
+                await StartWriteAsync(FrameHeader.Size).ConfigureAwait(false);
+                WriteFrameHeader(new FrameHeader(0, FrameType.Settings, FrameFlags.Ack, 0));
 
-            FinishWrite(mustFlush: true);
+                FinishWrite(mustFlush: true);
+            }
+            catch (Exception e)
```

Comment:
Why can't we handle this in FinishWrite, instead of always having to do try/catch in the callers? Is the exception coming from somewhere else? Seems like the only exception that should matter here (and elsewhere) is an exception from FinishWrite.

Reply:
how would you exactly envision that @geoffkizer? 
Righty now I only updated few functions which are always "ignored" but in normal case the  exception  is just thrown up the call stack. Also right now we would observe exception when trying to send reset but failure to do so, would not cause exception on  it's own - if for example server is already gone.

Reply:
Well, it seems to me we should be handling this in FinishWrite already. After all, the actual write may occur asynchronously from this code.

Let me ask the question this way: Why is this try/catch needed? You added it for some reason. Is it just to ensure that no exception escapes here? Why is that necessary?

Reply:
in previous change @stephentoub raised concern about UnobservedTaskExceptions. So I look for any background task which can fail and I tried to either observe the exception (and log it) or propagate it up.

It seems like in past UnobservedTaskExceptions would kill the process. Now it won't, but we are still trying to avoid it. So I tried to make sure any exception in "ignored" task can be observed.

Reply:
Some related observations here:

It's imperative that there's no exception between StartWriteAsync and FinishWrite in any of these routines. If that happened then the write lock would never be released.

So in theory, the only thing that can throw here is StartWriteAsync or FinishWrite. But in practice, I don't think FinishWrite actually ever throws either -- the write may happen asynchronously, so there's no real point in surfacing exceptions here. And StartWriteAsync should really only throw in the connection is already aborted.

So again, I'm not sure what we gain by having try/catch here...

Reply:
ok, I'll take another look. If that is base mode of operation, maybe I can add note to make it clear.
I did not come up with any good test case for this so it was mostly precaution for cases @stephentoub was concerned about.

On that note, how failures in write would be propagated up?

Reply:
The StartWriteAsync () calls  AcquireWriteLockAsync()

```c#
  if (_disposed)
  {
     throw new IOException(SR.net_http_request_aborted);
  }
```
we have seen this in unit tests and it seems like this can be way how background task gets to faulted state. In this case there is not much to do as the connection is already being aborted but I think it will lead to condition @stephentoub  was concerned about.

Reply:
I'm still not clear why we need all these new try/catch blocks.

StartWriteAsync can throw, but if so, it will release the writer lock.

Nothing between StartWriteAsync and FinishWrite should throw.

If FinishWrite throws, it should handle whatever cleanup is necessary, e.g. call Abort().

Reply:
Is the whole point here just to avoid UnobservedTaskExceptions? Is this really worthwhile? @stephentoub 

If it's really worth doing, then instead of modifying all these routines like SendSettingsAckAsync to never throw, we should simply add some sort of helper in the *caller* routines to eat the exception someway.

The point here is that all the Send*Async methods should have consistent behavior, and that behavior is to throw if the connection has been aborted (in StartWriteAsync) before we get a chance to grab the write lock. Some callers of these routines will care about this. Some won't. The callers themselves should decide whether they care (by awaiting) or not (by explicitly eating the exception).

Reply:
In fact, you are already doing this in a bunch of places. You're calling IgnoreExceptionsAsync on the Task returned from SendRstStreamAsync. Why can't we just do this other places as well?

Reply:
> Is the whole point here just to avoid UnobservedTaskExceptions? Is this really worthwhile? 

Ideally.  Developers do subscribe to this event in order to find places that they've allowed exceptions to go unhandled, and we hear from them when our exceptions clutter up their results.  It's also possible (though unusual) for a developer to change a setting that causes the app to crash for such unobserved exceptions (that was the default behavior in .NET Framework 4 until we switched it to not crash in .NET Framework 4.5 when async/await was introduced.)

That said, if suppressing it adds measurable cost or complexity, it's likely not worth it, at least not until we hear about it.

Reply:
> That said, if suppressing it adds measurable cost or complexity, it's likely not worth it, at least not until we hear about it.

I'm fine if we just add IgnoreExceptionsAsync at the calling sites. That seems to be what we do in other situations, e.g. similar cases for HTTP/1.1.

What I don't want to do is change the semantics of the actual Send*Async routines, which is what the current change does.

Reply:
That will be in next update @geoffkizer. I felt that is better than pollute code with the try/catch. As you mentioned reasons for failures should be already propagate from connection IO.